### PR TITLE
ec2: Allow EC2-VPC instances to modify security groups - fixes #18928

### LINKED ISSUE #18928: Alter Security Groups on EC2 Instances
_From @dbrgn on December 5, 2014 15:47_

##### Issue Type:

Feature Idea
##### Component Name:

ec2 module
##### Ansible Version:

ansible 1.8.1
##### Environment:

N/A
##### Summary:

I did not find a way to change EC2 Security Groups on an existing instance. It is possible to define the groups when creating a new instance with the "ec2" module, but nto to alter, add or remove groups on an existing instance.

Rationale: I create an instance with a limited set of security groups, so that the software cannot connect to the server yet during provisioning. Then I provision the host, and at the end I want to add a security group so that the host can start connecting to the server and process work.
##### Steps To Reproduce:

N/A
##### Expected Results:

Being able to:
- Add security groups to an existing EC2 instance
- Remove security groups from an existing EC2 instance
- Overwrite the security groups on an existing EC2 instance
##### Actual Results:

N/A

_Copied from original issue: ansible/ansible-modules-core#454_

### PR DESCRIPTION
**Pull request by s-hertel**
make ec2 pep8

##### SUMMARY
EC2-VPC instances can modify security groups. Added that feature to the ec2 module. Made ec2 pep8 and removed from legacy file. Fixes #18928

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/ec2.py

##### ANSIBLE VERSION
```
2.4.0
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/ec2.py [hunk 1]
```diff
@@ -1438,6 +1440,24 @@ def startstop_instances(module, ec2, instance_ids, state, instance_tags):
                 inst.modify_attribute('disableApiTermination', termination_protection)
                 changed = True
 
+            # Check security groups and if we're using ec2-vpc; ec2-classic security groups may not be modified
+            if inst.vpc_id and (group_id or group_name):
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2.py]: Can you restructure this to:

```
if inst.vpc_id and group_name:
    # current contents of the if group_name$s
elif inst.vpc_id and group_id:
    # the contents of the elif below
```

To bring things in an indentation level?

**Review by ryansb — CHANGES_REQUESTED** [review]

**Review by ryansb — APPROVED** [review]
Looks great, thanks!

File: lib/ansible/modules/cloud/amazon/ec2.py [hunk 2]
```diff
@@ -1438,6 +1441,24 @@ def startstop_instances(module, ec2, instance_ids, state, instance_tags):
                 inst.modify_attribute('disableApiTermination', termination_protection)
                 changed = True
 
+            # Check security groups and if we're using ec2-vpc; ec2-classic security groups may not be modified
+            if inst.vpc_id and group_name:
+                grp_details = ec2.get_all_security_groups(filters={'vpc_id': inst.vpc_id})
+                if isinstance(group_name, string_types):
+                    group_name = [group_name]
+                unmatched = set(group_name) - set(to_text(grp.name) for grp in grp_details)
+                if len(unmatched) > 0:
```
Comment (abadger) [inline lib/ansible/modules/cloud/amazon/ec2.py]: note: tests for zero length and nonzero length containers can test the container directly:
``` python
# equivalent:
if unmatched: pass
if len(unmatched) == 0: pass

# also equivalent:
if not unmatched: pass
if len(unmatched) > 0: pass
```

**Review by ryansb — APPROVED** [review]

**Comment by davidfischer-ch** [comment]
Hello, is it possible to provide us with an example call ? I am eager to test it when released.

Thanks.

**Comment by s-hertel** [comment]
hi @davidfischer-ch, thanks for the interest.
This change allows you to run something like:
```
      ec2:
        region: us-east-1
        instance_ids: "{{ instance_ids }}"
        vpc_subnet_id: subnet-XXXXXXXX
        state: running
        group_id:
          - sg-XXXXXXXX
          - sg-XXXXXXXX
```
which replaces the instances' security groups with the newly specified one(s), and if you don't specify new groups (using `group_id` or `group`) no change will be made.

**Comment by AnishAnil** [comment]
What version of Ansible will this fix be available??
This is one of the most needed feature in AWS and hoping to have this released on priority.

**Comment by s-hertel** [comment]
@AnishAnil This fix is available in version 2.4.

**Pull request merged**